text/2020-09-08-easily-send-http-request-on-workflow-task.md

```diff
@@ -0,0 +1,228 @@
+
+We want store the state into one or several `annotation` of templates, but it
+could not be done in short future. There are no embedded `metav1.ObjectMeta`
+inside the `Template`. We'd better to treat `Template` with `WorkflowNode` as
+`PodTemplateSpec` to `Pod`. I think we should split CRDs for workflow into
+another subgroup, then upgrade to `v1alpha2`.
+
+If we want to show the form of configuration of HTTP request in frontend, we
+could only parse the command line of `curl`, and that is awful: it's hard to
+keep consistently between original parameters and parsed parameters. I prefer
+NOT to do that. But this feature is very important for user to modifying the
+exists workflow. So I have to implement this. Welcome to other way to implement
+this.
+
+The generated `Task` should show the preview instantly.
+
+### New context variable: `http`
+
+Although we could use `stdout` to collect the output of `curl`, it meaningless
+that we do not provide the parsed HTTP context for the next conditional
+branches.
+
+So we will provide another context variable called `http` for introducing the
+parsed HTTP context. It contains very simplified and common-used HTTP request
+and response, easy for being used in the conditions.
+
+#### Structure of `http`
+
+```golang
+type ContextVariableHTTP struct {
+    Request  HTTPRequest  `json:"request"`
+    Response HTTPResponse `json:"response"`
+}
+type HTTPRequest struct {
+    Method string      `json:"method"`
+    URL    string      `json:"url"`
+    Header http.Header `json:"header"`
+    Body   []byte      `json:"body"`
+}
+type HTTPResponse struct {
+    StatusCode int         `json:"statusCode"`
+    Header     http.Header `json:"header"`
+    Body       []byte      `json:"body"`
+}
+
+```
+
+#### Example conditions
+
+Response returns 20x:
+
+`http.response.statusCode >= 200 && http.response.statusCode < 300`
+
+Response returns 50x:
+
+`http.response.statusCode >= 500`
+
+And used could select these fields in context variable `http`, and use it in the
+condition.
+
+## Drawbacks
+
+There are several things that bother me a long time, please leave comments if
+you have any idea!
+
+### parsing curl command inline back into the parameters
+
+Here is one concerned thing about "parsing curl command inline back into the
+parameters".
+
+Origin demand:
+
+- restore the parameters of HTTP request, for the next show and modification on
+  workflow view.
+- restore the parameters of HTTP request, for the context variable
+  `http.request`
+
+Excepted implementation:
+
+Store all the parameters with `annotation`, with specific prefix like
+`chaos-mesh.opg/workflow.http.method=GET`
+
+Actual implementations:
+
+Because struct `Template` in WorkflowSpec does not embed `metav1.Object`, there
+is nowhere to place the annotations. So we have to parse the rendered "curl
+command line". We need a lot of works and tests for keep the parsed result is
+consistent with the original parameters.
```

Comment:
I have a question, why not just use go http client to do such thing. Yes, It's another reinvented wheel. But I think it's acceptable.

From the draft PR, I could see it's implemented by a function [ParseCommands](https://github.com/chaos-mesh/chaos-mesh/pull/2308/files#diff-f7fc68ac2b7dd2e5e2a7eabe888a113966ccf3dd0620678f888447b98453702eR26), and it rely on cli conversion which is little bit hack, I'm afraid that it would grow more complex and eventually hard to maintain.

For an example based on go http client, see blackbox exporter [ProbeHTTP](https://github.com/prometheus/blackbox_exporter/blob/master/prober/http.go#L235). We're actually planning to do similar thing, send request according the spec, and collect the result.

In a word, I prefer purge go implementation, it's flexible and easier to maintain, what do you think?

Reply:
The basic idea is reusing `Task`, and `Task` could only take commands.  🤔

I think your suggestion is mostly like "Alternative solution 1: New type of WorkflowNode/Template for sending HTTP request", it would introduce more types of `Template` and `WorkflowNode`.

Reply:
IMO, we have not designed any feature about the health-checking for the injected application, making a health-probe in the workflow would only bring some time-series data on the surface, and no one(even user) could not realize what does it means. 🙈

Based on that, sending HTTP requests is only a way to configure the behavior of Workflow for users who already know the key metrics of the application, or post a message to the external system.

Reply:
> From the draft PR, I could see it's implemented by a function ParseCommands, and it rely on cli conversion which is little bit hack, I'm afraid that it would grow more complex and eventually hard to maintain.

`ParseCommands` is just a temporary solution, I really want to restore these origin request parameters into the `annotations`. https://github.com/strrl/rfcs/blob/http-request-on-workflow-task/text/2020-09-08-easily-send-http-request-on-workflow-task.md#frontend-component-for-this-type-of-task-and-preview-of-generated-result 

The problem is we did not put `annotation`/`metav1.ObjectMeta` into the `Template`, so we decide to use `PraseCommands` for a while. The update of CRD should finished before another feature called "execute scripts in Task".

Reply:
> Based on that, sending HTTP requests is only a way to configure the behavior of Workflow for users who already know the key metrics of the application, or post a message to the external system.

Can't agree more. The http request is only used for confirming the application real state by interacting with the external systems, e.g. lb, service discovery, monitoring system etc.

Could you give me an example by this design ? If a user want to send HTTP request within the workflow task, what does the spec look like?

Reply:
> Could you give me an example by this design ?

Users could post messages to the monitoring system, like alertmanager, or trigger some action by webhook. And user could still use this for the health probe, with an existed `/health` HTTP endpoint on the application.

> If a user want to send HTTP request within the workflow task, what does the spec look like?

When user want to create HTTP request with yaml and kubectl, there is no changes: user still need handwrite commands and specify a docker image.
This RFC aims to improve the experience on Chaos dashboard, user could create an HTTP Task by filing a simple form: 
![image](https://user-images.githubusercontent.com/20221408/134867198-e8ddf10d-1c5d-494c-83af-e5f730a935f6.png)

Reply:
> When user want to create HTTP request with yaml and kubectl, there is no changes: user still need handwrite commands and specify a docker image.

So this RFC is just a handy wrapper for dashboard use, our users still could use their own image and do http request by themselves way, right?

Reply:
Yes!❤️